File: c/step5.c
#include <stdio.h>
#include <SDL2/SDL.h>
#include <SDL2/SDL_timer.h>
#include <SDL2/SDL_image.h>

#define TITLE "My SDL2 Window"
#define WINDOW_WIDTH 640
#define WINDOW_HEIGHT 480
#define SPEED 300

int main() {
    if (SDL_Init(SDL_INIT_VIDEO|SDL_INIT_TIMER) != 0) {
        printf("Error initializing SDL: %s\n", SDL_GetError());
        return 1;
    }

    SDL_Window *window = SDL_CreateWindow(TITLE,
                                          SDL_WINDOWPOS_CENTERED,
                                          SDL_WINDOWPOS_CENTERED,
                                          WINDOW_WIDTH,
                                          WINDOW_HEIGHT,
                                          0);

    if (!window) {
        printf("Error creating window: %s\n", SDL_GetError());
        SDL_Quit();
        return 1;
    }

    Uint32 render_flags = SDL_RENDERER_ACCELERATED | SDL_RENDERER_PRESENTVSYNC;
    
    SDL_Renderer *rend = SDL_CreateRenderer(window,
                                            -1,
                                            render_flags);

    if (!rend) {
        printf("Error creating renderer: %s\n", SDL_GetError());
        SDL_DestroyWindow(window);
        SDL_Quit();
        return 1;
    }

    SDL_Surface *surface = IMG_Load("imgs/newyearalien2.png");

    if (!surface) {
        printf("Error creating surface\n");
        SDL_DestroyRenderer(rend);
        SDL_DestroyWindow(window);
        SDL_Quit();
        return 1;
    }

    SDL_Texture *tex = SDL_CreateTextureFromSurface(rend, surface);
    SDL_FreeSurface(surface);

    if (!tex) {
        printf("Error creating texture: %s\n", SDL_GetError());
        SDL_DestroyRenderer(rend);
        SDL_DestroyWindow(window);
        SDL_Quit();
        return 1;
    }

    SDL_Rect dst;

    SDL_QueryTexture(tex, NULL, NULL, &dst.w, &dst.h);
    dst.w /= 4;
    dst.h /= 4;
    
    float x_pos = (WINDOW_WIDTH - dst.w) / 2;
    float y_pos = (WINDOW_HEIGHT -dst.h) / 2;

    float x_vel = SPEED;
    float y_vel = SPEED;

    int close_requested = 0;

    while (!close_requested) {

        SDL_Event event;

        while (SDL_PollEvent(&event)) {
            if (event.type == SDL_QUIT) {
                close_requested = 1;
            }
        }
        if (x_pos <= 0) {
            x_pos = 0;
            x_vel = -x_vel;
        }
        if (y_pos <= 0) {
            y_pos = 0;
            y_vel = -y_vel;
        }
        if (x_pos >= WINDOW_WIDTH - dst.w) {
            x_pos = WINDOW_WIDTH - dst.w;
            x_vel = -x_vel;
        }
        if (y_pos >= WINDOW_HEIGHT - dst.h) {
            y_pos = WINDOW_HEIGHT - dst.h;
            y_vel = -y_vel;
        }

        x_pos += x_vel / 60;
        y_pos += y_vel / 60;

        dst.y = (int) y_pos;
        dst.x = (int) x_pos;

        SDL_RenderClear(rend);

        SDL_RenderCopy(rend, tex, NULL, &dst);
        SDL_RenderPresent(rend);

        SDL_Delay(1000 / 60);
    }

    // Clean up
    SDL_DestroyTexture(tex);
    SDL_DestroyRenderer(rend);
    SDL_DestroyWindow(window);
    SDL_Quit();

    return 0;
}

File: nim/step5.nim
import sdl2/sdl,
       sdl2/sdl_image as img

const
  TITLE = "Nim SDL2 Window"
  WINDOW_WIDTH = 640
  WINDOW_HEIGHT = 480
  SPEED = 300

proc main() =
  var
    window: Window
    rend: Renderer
    surface: Surface
    tex: Texture

  defer:
    tex.destroyTexture()
    rend.destroyRenderer()
    window.destroyWindow()
    sdl.quit()

  if sdl.init(INIT_VIDEO or INIT_TIMER) != 0:
    echo "Error: Couldn't initialize SDL: ", sdl.getError()

  window = sdl.createWindow(TITLE,
                            WINDOWPOS_CENTERED,
                            WINDOWPOS_CENTERED,
                            WINDOW_WIDTH,
                            WINDOW_HEIGHT,
                            WINDOW_SHOWN)

  if window.isNil:
    echo "ERROR: Couldn't create window: ", sdl.getError()
    sdl.quit()

  rend = sdl.createRenderer(window,
                            -1,
                            RENDERER_ACCELERATED)

  if rend.isNil:
    echo "ERROR: Couldn't create renderer: ", sdl.getError()
    window.destroyWindow()
    sdl.quit()

  surface = img.load("../imgs/newyearalien2.png")

  if surface.isNil:
    echo "ERROR: Couldn't create surface: ", sdl.getError()
    rend.destroyRenderer()
    window.destroyWindow()
    sdl.quit()

  tex = sdl.createTextureFromSurface(rend, surface)
  surface.freeSurface()

  if tex.isNil:
    echo "ERROR: Couldn't create texture: ", sdl.getError()
    rend.destroyRenderer()
    window.destroyWindow()
    sdl.quit()

  var
    dest: sdl.Rect
    event: sdl.Event
    x_pos: float
    y_pos: float
    x_vel: float
    y_vel: float
    close_requested = true

  if tex.queryTexture(nil, nil, addr(dest.w), addr(dest.h)) != 0:
    echo "ERROR: Couldn't query texture: ", sdl.getError()
    tex.destroyTexture()
    window.destroyWindow()
    sdl.quit()
  
  dest.w = int(dest.w / 4)
  dest.h = int(dest.h / 4)
  x_vel = float(SPEED)
  y_vel = float(SPEED)

  while close_requested:

    while sdl.pollEvent(addr(event)) != 0:
      if event.kind == EventKind.QUIT:
        close_requested = false
        break

    if int(x_pos) <= 0:
      x_pos = float(0)
      x_vel = -x_vel

    if int(y_pos) <= 0:
      y_pos = float(0)
      y_vel = -y_vel

    if int(x_pos) >= WINDOW_WIDTH - dest.w:
      x_pos = float(WINDOW_WIDTH - dest.w)
      x_vel = -x_vel

    if int(y_pos) >= WINDOW_HEIGHT - dest.h:
      y_pos = float(WINDOW_HEIGHT - dest.h)
      y_vel = -y_vel

    x_pos += x_vel / 60
    y_pos += y_vel / 60

    dest.y = int(y_pos)
    dest.x = int(x_pos)

    if rend.renderClear() != 0:
      break

    if rend.renderCopy(tex, nil, addr(dest)) != 0:
      break

    rend.renderPresent()
    
    sdl.delay(uint32(1000/60))


when isMainModule:
  main()
